os_migrate/plugins/module_utils/server.py

```diff
@@ -248,3 +248,24 @@ def _refs_from_ser(self, conn, filters=None):
         refs['volumes'] = params['volumes']
 
         return refs
+
+    def dst_prerequisites_errors(self, conn, filters=None):
+        """Get messages for unmet destination cloud prerequisites including keypairs.
+
+        Returns: list of strings (error messages), empty when prerequisites
+        are met
+        """
+        errors = super().dst_prerequisites_errors(conn, filters)
+
+        # Parent class dst_prerequisites_errors does not validate keypairs exist in destination cloud.
+        # To do this we do a lookup to see if keypair name exists and append to errors list if not
+        params = self.params()
+        try:
+            # find_keypair only accepts user_id as a filter. Check if user_id in filters dict else use None
+            user_id = filters.get('user_id', None)
+            conn.compute.find_keypair(params['key_name'], ignore_missing=False, user_id=user_id)
```

Comment:
Not sure if providing `user_id=None` is the same as not providing it at all, but we'll see in the e2e tests i think.

More generally, unless i'm mistaken, we actually stopped using the filters for destination-side lookups as the references carry their own project/domain info nowadays. We just pass around the destination filters because they were used previously. Later we may want to check whether it's indeed the case, and remove the filters from destination-side lookups everywhere.

This is just for context though. If the CI passes, i think we can merge like this.

Reply:
13->16 is green

Reply:
Ack.

The default for user_id kwarg for find_keypair is None and it is checked

`attrs = {'user_id': user_id} if user_id else {}`

Reply:
it appears that there are versions of openstacksdk which do not accept user_id 

`find_keypair() got an unexpected keyword argument 'user_id'` from https://gitlab.cee.redhat.com/os-migrate/bot/-/jobs/4416270

Removing user_id as you suggested.

Reply:
Looks good.